tests/functional/sources/test_source_fresher_state.py

```diff
@@ -0,0 +1,727 @@
+import os
+import json
+import shutil
+import pytest
+from datetime import datetime, timedelta
+
+from dbt.exceptions import InternalException
+
+
+import dbt.version
+from tests.functional.sources.common_source_setup import BaseSourcesTest
+
+from tests.functional.sources.fixtures import (
+    error_models__schema_yml,
+    models__newly_added_model_sql,
+    models__newly_added_error_model_sql,
+)
+
+
+# TODO: We may create utility classes to handle reusable fixtures.
+def copy_to_previous_state():
+    shutil.copyfile("target/manifest.json", "previous_state/manifest.json")
+    shutil.copyfile("target/run_results.json", "previous_state/run_results.json")
+
+
+# put these here for now to get tests working
+class AnyStringWith:
+    def __init__(self, contains=None):
+        self.contains = contains
+
+    def __eq__(self, other):
+        if not isinstance(other, str):
+            return False
+
+        if self.contains is None:
+            return True
+
+        return self.contains in other
+
+    def __repr__(self):
+        return "AnyStringWith<{!r}>".format(self.contains)
+
+
+class AnyFloat:
+    """Any float. Use this in assertEqual() calls to assert that it is a float."""
+
+    def __eq__(self, other):
+        return isinstance(other, float)
+
+
+class SuccessfulSourceFreshnessTest(BaseSourcesTest):
+    @pytest.fixture(scope="class", autouse=True)
+    def setUp(self, project):
+        self.run_dbt_with_vars(project, ["seed"])
+        pytest._id = 101
+        pytest.freshness_start_time = datetime.utcnow()
+        # this is the db initial value
+        pytest.last_inserted_time = "2016-09-19T14:45:51+00:00"
+
+        os.environ["DBT_ENV_CUSTOM_ENV_key"] = "value"
+
+        yield
+
+        del os.environ["DBT_ENV_CUSTOM_ENV_key"]
+
+    def _set_updated_at_to(self, project, delta):
+        insert_time = datetime.utcnow() + delta
+        timestr = insert_time.strftime("%Y-%m-%d %H:%M:%S")
+        # favorite_color,id,first_name,email,ip_address,updated_at
+        insert_id = pytest._id
+        pytest._id += 1
+        quoted_columns = ",".join(
+            project.adapter.quote(c)
+            for c in ("favorite_color", "id", "first_name", "email", "ip_address", "updated_at")
+        )
+        kwargs = {
+            "schema": project.test_schema,
+            "time": timestr,
+            "id": insert_id,
+            "source": project.adapter.quote("source"),
+            "quoted_columns": quoted_columns,
+        }
+        raw_sql = """INSERT INTO {schema}.{source}
```

Comment:
Note for future reader: This probably isn't going to fly on other databases, but that's okay, the functionality being tested here / added in this PR is node selection, not adapter-specific